napari/layers/_tests/test_layer_attributes.py

```diff
@@ -18,3 +19,24 @@ def test_layer_bounding_box_order(image_shape, dims_displayed, expected):
     assert np.allclose(
         layer._display_bounding_box(dims_displayed=dims_displayed), expected
     )
+
+
+@pytest.mark.parametrize('Layer, data, ndim', layer_test_data)
+def test_layer_extent_cache(Layer, data, ndim):
+    np.random.seed(0)
+    layer = Layer(data)
+    # Check layer has been correctly created
+    assert layer.ndim == ndim
+
+    # Check layer extent doesn't change when unrelated attribute changes
+    layer.opacity = 1
+    old_extent = layer.extent
+    assert old_extent is layer.extent
+    layer.opacity = 0.5
+    assert old_extent is layer.extent
+
+    # Check layer extent change when scale changes
+    layer.scale = (2,) + layer.scale[1:]
+    old_extent2 = layer.extent
+    assert old_extent is not layer.extent
```

Comment:
I still don't love these assertions because they're testing an implementation detail that could change. This PR is an optimization that is intended not to change the functional behavior of napari. The tests act as a defense to make sure that functional behavior is maintained regardless of how we choose to implement how `layer.extent` gives us the right value.

Ideally, I think we'd assert that the extent has the expected value based on the attribute we update (e.g. scale), but that's not easy

The test name at least does reflect that this is testing the cache as an implementation detail, so I don't think it's a blocker.

Reply:
100% agreed, i've tried to improve, and we can merge as is now and then update later or we can change now. i really don't mind

Reply:
I like the changes and I think these are good enough to merge now. We can think about better coverage through parameterization later.